File: old/test/proto/cs/sample1.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Net;

namespace NodeProtocol {

	public class Sample1 {

		// public constants
		public const uint ID = 0;
		public const string NAME = "sample1";
		// public properties
		public uint Id { get; set; }
		public string Key { get; set; }
		public string Value { get; set; }
		public bool Enabled { get; set; }
		public NodeProtocol.Sample2[] Sample2list { get; set; }
		public sbyte _eight { get; set; }
		public short _sixteen { get; set; }
		public int _thirtytwo { get; set; }
		public System.DateTime Datetime { get; set; }

		// private properties
		private byte[] _4bytes = new byte[4];
		private byte[] _2bytes = new byte[2];
		private byte[] _8bytes = new byte[8];


		public Sample1(byte[] buf = null) {
			if (buf != null) {
				int offset = 0;
				
				// Unique ID: unpack Uint32 as Id
				Buffer.BlockCopy(buf, offset, _4bytes, 0, 4);
				Array.Reverse(_4bytes);
				Id = (uint)BitConverter.ToUInt32(_4bytes, 0);
				offset += 4;
				// unpack string as Key
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				Array.Reverse(_2bytes);
				offset += 2;
				int keySize = (int)BitConverter.ToUInt16(_2bytes, 0);
				byte[] keyBytes = new byte[keySize];
				Buffer.BlockCopy(buf, offset, keyBytes, 0, keySize);
				offset += keySize;
				Key = Encoding.UTF8.GetString(keyBytes);
				// Probably JSON: unpack string as Value
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				Array.Reverse(_2bytes);
				offset += 2;
				int valueSize = (int)BitConverter.ToUInt16(_2bytes, 0);
				byte[] valueBytes = new byte[valueSize];
				Buffer.BlockCopy(buf, offset, valueBytes, 0, valueSize);
				offset += valueSize;
				Value = Encoding.UTF8.GetString(valueBytes);
				// unpack bool as Enabled
				bool enabledBool = false;
				byte enabledValue = (byte)buf[offset];
				if (enabledValue == 1) {
					enabledBool = true;
				}
				Enabled = enabledBool;
				offset += 1;
				// unpack sample2list as array of sample2
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				Array.Reverse(_2bytes);
				offset += 2;
				int sample2listLength = (int)BitConverter.ToUInt16(_2bytes, 0);
				Sample2list = new Sample2[sample2listLength];
				int sample2listIndex = 0;
				for (sample2listIndex = 0; sample2listIndex < sample2listLength; sample2listIndex++) {
					// unpack Sample2 as Sample2list
					Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
					Array.Reverse(_2bytes);
					int sample2listSize = (int)BitConverter.ToUInt16(_2bytes, 0);
					offset += 2;
					byte[] sample2listBytes = new byte[sample2listSize];
					Buffer.BlockCopy(buf, offset, sample2listBytes, 0, sample2listSize);
					Sample2list[sample2listIndex] = new Sample2(sample2listBytes);
					offset += sample2listSize;
				}
				// unpack sbyte as _eight
				_eight = (sbyte)buf[offset];
				offset++;
				// unpack short as _sixteen
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				_sixteen = (short)BitConverter.ToInt16(_2bytes, 0);
				offset += 2;
				// unpack int as _thirtytwo
				Buffer.BlockCopy(buf, offset, _4bytes, 0, 4);
				_thirtytwo = (int)BitConverter.ToInt32(_4bytes, 0);
				offset += 4;
				// Date time object: unpack System.DateTime as Datetime
				Buffer.BlockCopy(buf, offset, _8bytes, 0, 8);
				Array.Reverse(_8bytes);
				TimeSpan datetimeTime = TimeSpan.FromMilliseconds(BitConverter.ToDouble(_8bytes, 0));
				Datetime = new DateTime(1970, 1, 1) +  datetimeTime;
				offset += 8;

			}
		}

		public byte[] Pack() {
			
			int totalBufSize = 0;
			// get byte size first: Unique ID: get byte size of Uint32 as Id
			byte[] idBytes = BitConverter.GetBytes(Id);
			Array.Reverse(idBytes);
			totalBufSize += 4;
			// get byte size first: get byte size of string as Key
			byte[] keyBytes = Encoding.UTF8.GetBytes(Key);
			totalBufSize += 2 + keyBytes.Length;
			// get byte size first: Probably JSON: get byte size of string as Value
			byte[] valueBytes = Encoding.UTF8.GetBytes(Value);
			totalBufSize += 2 + valueBytes.Length;
			// get byte size first: get byte size of bool as Enabled
			int enabledBool = 0;
			if (Enabled == true) {
				enabledBool = 1;
			}
			byte[] enabledBytes = BitConverter.GetBytes(enabledBool);
			totalBufSize += 1;
			// Sample2list is an array of Sample2
			totalBufSize += 2;
			List<byte[]> sample2listPackByteList = new List<byte[]>();
			for (int sample2listIndex = 0, sample2listArrayLength = Sample2list.Length; sample2listIndex < sample2listArrayLength; sample2listIndex++) {
				// get byte size first: get byte size of Sample2 as Sample2list
				/*** Sample2list ***/
				// get byte size first: get byte size of string as Sample2list[sample2listIndex].Name
				byte[] nameBytes = Encoding.UTF8.GetBytes(Sample2list[sample2listIndex].Name);
				totalBufSize += 2 + nameBytes.Length;

				// create pack buffer
				byte[] buf = new byte[totalBufSize];
				int offset = 0;

				// pack string as Name
				_2bytes = BitConverter.GetBytes((ushort)nameBytes.Length);
				Array.Reverse(_2bytes);
				Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
				offset += 2;
				Buffer.BlockCopy(nameBytes, 0, buf, offset, nameBytes.Length);
				sample2listPackByteList.Add(sample2listBytes);
			}
			// get byte size first: get byte size of sbyte as _eight
			byte[] _eightBytes = BitConverter.GetBytes(_eight);
			totalBufSize += 1;
			// get byte size first: get byte size of short as _sixteen
			byte[] _sixteenBytes = BitConverter.GetBytes(_sixteen);
			totalBufSize += 2;
			// get byte size first: get byte size of int as _thirtytwo
			byte[] _thirtytwoBytes = BitConverter.GetBytes(_thirtytwo);
			totalBufSize += 4;
			// get byte size first: Date time object: get byte size of System.DateTime as Datetime
			double datetimeTime = (Datetime - new DateTime(1970, 1, 1)).TotalMilliseconds;
			byte[] datetimeBytes = BitConverter.GetBytes((double)datetimeTime);
			Array.Reverse(datetimeBytes);
			totalBufSize += 8;

			// create pack buffer
			byte[] buf = new byte[totalBufSize];
			int offset = 0;

			// Unique ID: pack Uint32 as Id
			Array.Reverse(idBytes);
			Buffer.BlockCopy(idBytes, 0, buf, offset, idBytes.Length);
			offset += 4;
			// pack string as Key
			_2bytes = BitConverter.GetBytes((ushort)keyBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(keyBytes, 0, buf, offset, keyBytes.Length);
			// Probably JSON: pack string as Value
			_2bytes = BitConverter.GetBytes((ushort)valueBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(valueBytes, 0, buf, offset, valueBytes.Length);
			// pack bool as Enabled
			Buffer.BlockCopy(enabledBytes, 0, buf, offset, enabledBytes.Length);
			offset += 1;
			// Sample2list is an array of Sample2
			_2bytes = BitConverter.GetBytes((ushort)Sample2list.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			for (int sample2listIndex = 0, sample2listArrayLength = Sample2list.Length; sample2listIndex < sample2listArrayLength; sample2listIndex++) {
				// pack Sample2 as Sample2list

				int totalBufSize = 0;
				// get byte size first: get byte size of string as Name
				byte[] nameBytes = Encoding.UTF8.GetBytes(Name);
				totalBufSize += 2 + nameBytes.Length;

				// create pack buffer
				byte[] buf = new byte[totalBufSize];
				int offset = 0;

				// pack string as Name
				_2bytes = BitConverter.GetBytes((ushort)nameBytes.Length);
				Array.Reverse(_2bytes);
				Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
				offset += 2;
				Buffer.BlockCopy(nameBytes, 0, buf, offset, nameBytes.Length);
			}
			// pack sbyte as _eight
			Buffer.BlockCopy(_eightBytes, 0, buf, offset, _eightBytes.Length);
			offset += 1;
			// pack short as _sixteen
			Buffer.BlockCopy(_sixteenBytes, 0, buf, offset, _sixteenBytes.Length);
			offset += 2;
			// pack int as _thirtytwo
			Buffer.BlockCopy(_thirtytwoBytes, 0, buf, offset, _thirtytwoBytes.Length);
			offset += 4;
			// Date time object: pack System.DateTime as Datetime
			Array.Reverse(datetimeBytes);
			Buffer.BlockCopy(datetimeBytes, 0, buf, offset, datetimeBytes.Length);
			offset += 8;

			return buf;
		}

	}
}

File: old/test/proto/js/sample1.js
'use strict';

const id = 0;
const name = 'sample1';

module.exports.getId = getId;
module.exports.name = getName;
module.exports.create = create;
module.exports.pack = pack;
module.exports.unpack = unpack;

function getId() {
	return id;
}

function getName() {
	return name;
}

function create() {
	return {
		id: 0,
		key: '',
		value: '',
		enabled: true,
		sample2list: [],
		_eight: 0,
		_sixteen: 0,
		_thirtytwo: 0,
		datetime: new Date()
	};
}

// converts an object into binary
function pack(obj) {
	const list = [];
	
	var idBytes = Bin.alloc(4);
	idBytes.writeUInt32BE(obj.id, 0);
	list.push(idBytes);
	var keyBytes = Bin.from(obj.key);
	var keySize = keyBytes.length;
	var keySizeBytes = Bin.alloc(2);
	keySizeBytes.writeUInt16BE(keySize, 0);
	list.push(keySizeBytes);
	list.push(keyBytes);
	var valueBytes = Bin.from(obj.value);
	var valueSize = valueBytes.length;
	var valueSizeBytes = Bin.alloc(2);
	valueSizeBytes.writeUInt16BE(valueSize, 0);
	list.push(valueSizeBytes);
	list.push(valueBytes);
	var enabledBytes = Bin.alloc(1);
	enabledBytes.writeUInt8(obj.enabled ? 0x01 : 0x00);
	list.push(enabledBytes);
	// .sample2list is an array
	const sample2listLengthBytes = Bin.alloc(2);
	sample2listLengthBytes.writeUInt16BE(obj.sample2list.length);
	list.push(sample2listLengthBytes);
	var sample2listIndex = 0;
	var sample2listLen = obj.sample2list.length;
	for (var sample2listIndex = 0; sample2listIndex < sample2listLen; sample2listIndex++) {

		var sample2listNameBytes = Bin.from(obj.sample2list[sample2listIndex].name);
		var sample2listNameSize = sample2listNameBytes.length;
		var sample2listNameSizeBytes = Bin.alloc(2);
		sample2listNameSizeBytes.writeUInt16BE(sample2listNameSize, 0);
		list.push(sample2listNameSizeBytes);
		list.push(sample2listNameBytes);
	}
	var _eightBytes = Bin.alloc(1);
	_eightBytes.writeInt8(obj._eight, 0);
	list.push(_eightBytes);
	var _sixteenBytes = Bin.alloc(2);
	_sixteenBytes.writeInt16LE(obj._sixteen, 0);
	list.push(_sixteenBytes);
	var _thirtytwoBytes = Bin.alloc(4);
	_thirtytwoBytes.writeInt32LE(obj._thirtytwo, 0);
	list.push(_thirtytwoBytes);
	var datetimeBytes = Bin.alloc(8);
	datetimeBytes.writeDoubleBE(obj.datetime.getTime(), 0);
	list.push(datetimeBytes);

	return Buffer.concat(list);
}

function unpack(buf) {
	const obj = {};
	
	var offset = 0;
	// Unique ID
	obj.id = buf.readUInt32BE(offset);
	offset += 4;
	// unpack for obj.key
	var keySize = buf.readUInt16BE(offset);
	offset += 2;
	var keyBytes = Bin.alloc(keySize);
	buf.copy(keyBytes, 0, offset);
	offset += keySize;
	obj.key = keyBytes.toString();
	// Probably JSON
	var valueSize = buf.readUInt16BE(offset);
	offset += 2;
	var valueBytes = Bin.alloc(valueSize);
	buf.copy(valueBytes, 0, offset);
	offset += valueSize;
	obj.value = valueBytes.toString();
	// unpack for obj.enabled
	obj.enabled = buf.readUInt8(offset) === 0x01 ? true : false;
	offset += 1;
	// unpack for obj.sample2list
	obj.sample2list = [];
	var sample2listLength = buf.readUInt16BE(offset);
	offset += 2;
	var sample2listLen = sample2listLength;
	for (var sample2listIndex = 0; sample2listIndex < sample2listLength; sample2listIndex++) {

		obj.sample2list[sample2listIndex] = {};
		// unpack for obj.sample2list.name
		var sample2listNameSize = buf.readUInt16BE(offset);
		offset += 2;
		var sample2listNameBytes = Bin.alloc(sample2listNameSize);
		buf.copy(sample2listNameBytes, 0, offset);
		offset += sample2listNameSize;
		obj.sample2list[sample2listIndex].name = sample2listNameBytes.toString();
	}
	// unpack for obj._eight
	obj._eight = buf.readInt8(offset);
	offset += 1;
	// unpack for obj._sixteen
	obj._sixteen = buf.readInt16LE(offset);
	offset += 2;
	// unpack for obj._thirtytwo
	obj._thirtytwo = buf.readInt32LE(offset);
	offset += 4;
	// Date time object
	obj.datetime = new Date(buf.readDoubleBE(offset));
	offset += 8;

	return obj;
}

// Buffer handler static object Bin: Bin.from(<string>) and Bin.alloc(<size>)
const __methods = {
	from: null,
	alloc: null
};

getNodeVersion();

const Bin = {
	from: __methods.from,
	alloc: __methods.alloc
};

function getNodeVersion() {
	const vstring = process.version.replace('v', '');
	const v = parseInt(vstring.substring(0, vstring.indexOf('.')));
	if (v >= 7) {
		// version 7 or higher
		__methods.from = from;
		__methods.alloc = alloc;
	} else {
		__methods.from = oldFrom;
		__methods.alloc = oldAlloc;
	}
}

function from(str, enc) {
	return Buffer.from(str, enc);
}

function alloc(size) {
	return Buffer.alloc(size).fill(0);
}

function oldFrom(str, enc) {
	return new Buffer(str, enc);
}

function oldAlloc(size) {
	const buf = new Buffer(size);
	buf.fill(0);
	return buf;
}

